### PR TITLE
Add agent replicator component

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/replicator/StreamDataReplicator.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import com.netflix.titus.common.runtime.TitusRuntime;
+import com.netflix.titus.common.util.rx.ObservableExt;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Observable;
+import rx.Subscription;
+import rx.subjects.PublishSubject;
+
+public abstract class StreamDataReplicator<D> implements DataReplicator<D> {
```
Comment (corindwyer) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/replicator/StreamDataReplicator.java]: add comment on what is this class and why it is needed.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/agent/replicator/GrpcAgentReplicatorEventStream.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                        instanceGroupsById.put(id, GrpcAgentModelConverters.toCoreAgentInstanceGroup(event.getInstanceGroupUpdate().getInstanceGroup()));
+                        break;
+                    case AGENTINSTANCEUPDATE:
+                        com.netflix.titus.api.agent.model.AgentInstance instance = GrpcAgentModelConverters.toCoreAgentInstance(event.getAgentInstanceUpdate().getInstance());
+                        instancesByGroupId.computeIfAbsent(instance.getInstanceGroupId(), gid -> new ArrayList<>()).add(instance);
+                        break;
+                }
+            });
+
+            // No longer needed
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/agent/replicator/GrpcAgentReplicatorEventStream.java]: Done

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/master/DefaultTitusMasterOperations.java [hunk 3]
```diff
@@ -19,18 +21,23 @@
...
+    public DefaultTitusMasterOperations(AgentDataReplicator agentDataReplicator,
+                                        JobDataReplicator jobDataReplicator) {
+        this.jobDataReplicator = jobDataReplicator;
 
         this.subscription = Observable.interval(0, 5, TimeUnit.SECONDS).subscribe(tick -> {
-            JobCache snapshot = jobCacheResolver.getCurrent();
-            logger.info("Job snapshot: jobs={}, tasks={}, latency={}", snapshot.getJobs().size(),
-                    snapshot.getTasks().size(), toTimeUnitString(jobCacheResolver.getStalenessMs()));
+            AgentSnapshot agentSnapshot = agentDataReplicator.getCurrent();
+            logger.info("Agent snapshot: instanceGroups={}, instances={}, latency={}", agentSnapshot.getInstanceGroups().size(),
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/master/DefaultTitusMasterOperations.java]: Fixed

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/master/DefaultTitusMasterOperations.java [hunk 4]
```diff
@@ -19,18 +21,23 @@
...
+                    agentSnapshot.getInstances().size(), toTimeUnitString(agentDataReplicator.getStalenessMs()));
+
+            JobSnapshot jobSnapshot = jobDataReplicator.getCurrent();
+            logger.info("Job snapshot: jobs={}, tasks={}, latency={}", jobSnapshot.getJobs().size(),
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/master/DefaultTitusMasterOperations.java]: Fixed

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/agent/AgentSnapshot.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+import static java.util.Collections.unmodifiableMap;
+
+public class AgentSnapshot {
+
+    private final Map<String, AgentInstanceGroup> instanceGroupsById;
+    private final List<AgentInstanceGroup> instanceGroupList;
+
+    private final Map<String, AgentInstance> instancesById;
+    private final List<AgentInstance> instances;
+    private final Map<String, List<AgentInstance>> instancesByGroupId;
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/agent/AgentSnapshot.java]: Fixed

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/replicator/StreamDataReplicator.java (hunk 1 above)
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/replicator/StreamDataReplicator.java]: Done

**Pull request merged**